gensim/models/tfidfmodel.py

```diff
@@ -390,7 +390,7 @@ def initialize(self, corpus):
         self.num_nnz = numnnz
         self.dfs = dfs
         # and finally compute the idf weights
-        n_features = max(dfs) if dfs else 0
+        n_features = len(dfs) if dfs else 0
```

Comment:
What will this do when `dfs = {1: 123, 10000: 124}`?

See the original line (before someone introduced this off-by-one bug in some refactoring) for the correct solution.

Reply:
Well for the case `dfs = {1: 123, 10000: 124}`, I thought that 2 should be the correct number of features since IDs are just numbers that are attached to the tokens.

Reply:
@piskvorky Using the original line `1 + max([-1] + dfs.keys())` results in a TypeError:
`TypeError: can only concatenate list (not "dict_keys") to list`
Should I just cast the keys to a python list?

Reply:
I see what you mean, but following such logic would lead to segfaults. Gensim relies on the fact that *all* features in a given feature range are valid, even if not present in a particular training corpus. They maybe be still present in a test / prediction-time corpus.

I'm strongly -1 on redefining the meaning of  "number of features" here, even if it's just for logging. It would lead to inconsistencies and ultimately bugs.

Reply:
I agree that this change might be confusing if other parts of gensim assume that missing token IDs are implicitly assigned to features yet to be defined.

Reply:
They aren't yet-to-be-defined. They are defined, but didn't appear in that particular corpus.